src/lightning/pytorch/trainer/trainer.py

```diff
@@ -1018,6 +1018,19 @@ def _teardown(self) -> None:
         self._logger_connector.teardown()
         self._signal_connector.teardown()
 
+    def teardown(self) -> None:
```

Comment:
The Trainer already has a protected `_teardown` hook that will teardown the loops when it finishes.

Reply:
Yes, I'm aware. But with the public `teardown`, there are no failing tests. Without it, there are around 40 failing tests due to references to the dataloader after the trainer is torn down.

Reply:
Just an update: Fixing many of the failing test cases is as straightforward as using the dataloader attached to the model instead of the trainer. Others are not so easy. For example, some objects have attributes/computed properties that depend on the number of dataloaders attached to it.

In my opinion, it is not worth having so many breaking changes when it can be fixed with none. These are my suggestions to close this PR and its referenced issue:
1. Declining the PR and manually applying the fix mentioned in the issue ticket (this can be done outside lightning or documented somewhere, and is how I'm going about things right now).
2. Keeping a public `teardown` (or any other suitable name) method that applies the fix.
(3). Keeping this PR/issue open and waiting for suggestions/inputs to go about fixing the failing tests (depends on whether the redesign/refactoring is worth the effort).

Reply:
in general, I would agree but not sure about adding this public API...
have you tried to if `del trainer` also releases all memory?

Reply:
Yes, `del trainer` would release all memory but then I'd have no trainer to test/make inferences. My initial goal when creating the issue was to be able to safely attach the trainer to my "Estimator" object (this is a step in my ML pipeline) using `self.trainer = trainer` and not have the whole dataloader get attached as well.